frontend/src/metabase-lib/lib/Dimension.ts

```diff
@@ -815,80 +818,123 @@ export class FieldDimension extends Dimension {
-        return this._createField(fieldObject, { hydrate: true });
-      }
+  _getFieldMetadataFromNestedCard() {
+    const identifierProp = this._getIdentifierProp();
+    const virtualTableCardId = getQuestionIdFromVirtualTableId(
+      this.query()?.sourceTableId?.(),
+    );
+    if (virtualTableCardId != null) {
+      const question = this._metadata?.question(virtualTableCardId);
+      const field = question
+        ? _.findWhere(question.getResultMetadata(), {
+            [identifierProp]: this.fieldIdOrName(),
+          })
+        : undefined;
 
-      if (fieldMetadata) {
-        return this._createField(fieldMetadata);
-      }
+      return field;
+    }
+  }
 
-      return this._createField({ id: this._fieldIdOrName });
-    }
-
-    // look for a "virtual" field on the query's table or question
-    // for example, fields from a question based on a nested question have fields
-    // that show up in a card's `result_metadata`
-    if (this.query()) {
-      const table = this.query().table();
-
-      if (table != null) {
-        const field = _.findWhere(table.fields, {
-          name: this.fieldIdOrName(),
-        });
-
-        if (field) {
-          if (!fieldMetadata || !shouldMergeFieldResultMetadata) {
-            return field;
-          }
-          const fieldObject = merge(
-            field instanceof Field ? field.getPlainObject() : field,
-            fieldMetadata,
-          );
-          return this._createField(fieldObject, { hydrate: true });
-        }
+  _getFieldMetadataFromQueryTable() {
+    const identifierProp = this._getIdentifierProp();
+    const table = this.query()?.table();
+    return _.findWhere(table?.fields, {
+      [identifierProp]: this.fieldIdOrName(),
+    });
+  }
+
+  _combineFieldWithExtraMetadata(field: Field | undefined, fieldMetadata) {
+    if (field) {
+      if (!fieldMetadata || field === fieldMetadata) {
+        return field;
       }
 
-      if (fieldMetadata) {
-        return this._createField(fieldMetadata);
+      const fieldObject = merge(
+        field instanceof Field ? field.getPlainObject() : field,
+        fieldMetadata instanceof Field
+          ? fieldMetadata.getPlainObject()
+          : fieldMetadata,
+      );
+      return this._createField(fieldObject, { hydrate: true });
+    }
+
+    if (fieldMetadata) {
+      if (fieldMetadata instanceof Field) {
+        return fieldMetadata;
       }
+
+      return this._createField(fieldMetadata);
     }
+  }
 
-    // despite being unable to find a field, we _might_ still have enough data to know a few things about it
-    // for example, if we have an mbql field reference, it might contain a `base-type`
+  field(): Field {
+    // If a Field is cached on the FieldDimension instance, we can shortwire this method and
+    // return the cached Field.
+    const cachedField = this._getTrustedFieldCachedOnInstance();
+    if (cachedField) {
+      return cachedField;
+    }
+
+    let fieldMetadata;
```

Comment:
The code for setting `fieldMetadata` is in part a workaround for a bug in how we cache fields. The `result_metadata` fields that come back from a query to the table endpoint where the table is `card__123` are cached like normal fields even though they have unique metadata defined on them. This forces us to rely on the card associated with a dimension because the fields are clobbered as soon as a GET goes out for a table that shares fields.